In [1]:
import numpy as np


In [2]:
import sys
from pathlib import Path

# Detect project root (whether you're in notebooks/ or root)
PROJECT_ROOT = Path.cwd().parent if Path.cwd().name == "notebooks" else Path.cwd()
SRC_PATH = PROJECT_ROOT / "src"

# Add src/ to sys.path if not already there
if str(SRC_PATH) not in sys.path:
    sys.path.append(str(SRC_PATH))

print("✅ src added to path:", SRC_PATH.exists())


✅ src added to path: True


In [3]:
import sys
sys.path.append("..")  # Go up one level to access src folder

from src.experiment import solve_bond_selection
print("✅ Imported successfully")


▶ src.experiment imported and running
✅ Imported successfully


In [4]:
import os, sys, pathlib

# 1) Set working directory to your repo root (where README and src/ live)
repo_root = pathlib.Path.cwd()
print("Working in:", repo_root)

# 2) Ensure Python can import from src/
if str(repo_root) not in sys.path:
    sys.path.append(str(repo_root))


Working in: C:\Users\Admin\Documents\WISER_Optimization_VG\notebooks


In [5]:
%cd C:\Users\Admin\Documents\WISER_Optimization_VG


C:\Users\Admin\Documents\WISER_Optimization_VG


C:\Users\Admin\Documents\WISER_Optimization_VG\.venv310\lib\site-packages\IPython\core\magics\osm.py:417: UserWarning: This is now an optional IPython functionality, setting dhist requires you to install the `pickleshare` library.
  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]


In [6]:
import sys
from pathlib import Path

# Detect project root (whether you're in notebooks/ or root)
PROJECT_ROOT = Path.cwd().parent if Path.cwd().name == "notebooks" else Path.cwd()
SRC_PATH = PROJECT_ROOT / "src"

# Add src/ to sys.path if not already there
if str(SRC_PATH) not in sys.path:
    sys.path.append(str(SRC_PATH))

print("✅ src added to path:", SRC_PATH.exists())


✅ src added to path: True


In [7]:
sys.path.append(str(PROJECT_ROOT))
from src.experiment import solve_bond_selection




In [8]:
import os, sys, pathlib

repo_root = pathlib.Path.cwd()
print("Working in:", repo_root)

if str(repo_root) not in sys.path:
    sys.path.append(str(repo_root))


Working in: C:\Users\Admin\Documents\WISER_Optimization_VG


In [9]:
import os, sys, pathlib

repo_root = pathlib.Path.cwd()
print("Working in:", repo_root)

if str(repo_root) not in sys.path:
    sys.path.append(str(repo_root))


Working in: C:\Users\Admin\Documents\WISER_Optimization_VG


In [10]:
from pathlib import Path
import os

# Get current folder
cwd = Path.cwd()
print("Jupyter is running in:", cwd)

# Go up one level if you're inside /notebooks
if cwd.name == "notebooks":
    os.chdir(cwd.parent)
    print("Switched to:", Path.cwd())


Jupyter is running in: C:\Users\Admin\Documents\WISER_Optimization_VG


In [11]:
import sys
PROJECT_ROOT = Path.cwd()

if str(PROJECT_ROOT) not in sys.path:
    sys.path.append(str(PROJECT_ROOT))

print("Root added to sys.path:", str(PROJECT_ROOT))


Root added to sys.path: C:\Users\Admin\Documents\WISER_Optimization_VG


In [12]:
def touch_init(path):
    path.mkdir(parents=True, exist_ok=True)
    init = path / "__init__.py"
    if not init.exists():
        init.touch()
        print("✅ Created __init__.py in", path)
    else:
        print("✔️ __init__.py already exists in", path)

# Apply to src/ and any nested package folders
for rel in ["src", "src/sbo", "src/sbo/src", "src/sbo/src/optimizer"]:
    folder = PROJECT_ROOT / rel
    if folder.exists():
        touch_init(folder)


✔️ __init__.py already exists in C:\Users\Admin\Documents\WISER_Optimization_VG\src
✔️ __init__.py already exists in C:\Users\Admin\Documents\WISER_Optimization_VG\src\sbo
✔️ __init__.py already exists in C:\Users\Admin\Documents\WISER_Optimization_VG\src\sbo\src
✔️ __init__.py already exists in C:\Users\Admin\Documents\WISER_Optimization_VG\src\sbo\src\optimizer


In [13]:
exp_path = PROJECT_ROOT / "src" / "experiment.py"
if not exp_path.exists():
    raise FileNotFoundError("Missing: src/experiment.py")

text = exp_path.read_text(encoding="utf-8")
replacements = {
    "from .sbo.src.optimizer.optimization_monitor": "from src.sbo.src.optimizer.optimization_monitor",
    "from .sbo.src.optimizer.optimization_wrapper": "from src.sbo.src.optimizer.optimization_wrapper",
}

patched = text
for old, new in replacements.items():
    if old in patched:
        patched = patched.replace(old, new)

if patched != text:
    backup = exp_path.with_suffix(".py.bak")
    backup.write_text(text, encoding="utf-8")
    exp_path.write_text(patched, encoding="utf-8")
    print("✅ Patched experiment.py (backup saved as experiment.py.bak)")
else:
    print("✔️ No changes needed — looks clean.")


✔️ No changes needed — looks clean.


In [14]:
from src.experiment import solve_bond_selection
print("✅ Import OK")


✅ Import OK


In [15]:
LP_PATH = "data/1/31bonds/docplex-bin-avgonly.lp"  # you used this file earlier


In [16]:
LP_REL = "data/1/31bonds/docplex-bin-avgonly.lp"
x, fx, dt = solve_bond_selection(LP_REL)
print("🔢 Bonds selected:", int(np.sum(x)))
print("📈 Objective value:", round(fx, 4))
print("⏱ Runtime (seconds):", round(dt, 2))


🔢 Bonds selected: 21
📈 Objective value: 40.2771
⏱ Runtime (seconds): 45.39


In [17]:
import os
print(os.getcwd())


C:\Users\Admin\Documents\WISER_Optimization_VG


In [18]:
import sys
sys.path.append("..")  # Moves up one level to access src/

from src.experiment import solve_bond_selection
print("✅ Imported successfully")


✅ Imported successfully


In [19]:
import numpy as np

LP_PATH = "data/1/31bonds/docplex-bin-avgonly.lp"  # Update if needed

x, fx, dt = solve_bond_selection(LP_PATH)

print("🔢 Bonds selected:", int(np.sum(x)))
print("📈 Objective value:", round(fx, 4))
print("⏱ Runtime (seconds):", round(dt, 2))


🔢 Bonds selected: 21
📈 Objective value: 40.2771
⏱ Runtime (seconds): 44.83


In [20]:
import os

for root, dirs, files in os.walk(".."):
    for name in files:
        if name.endswith(".lp"):
            print(os.path.join(root, name))


..\WISER_Optimization_VG\data\1\31bonds\docplex-bin-avgonly-nocplexvars.lp
..\WISER_Optimization_VG\data\1\31bonds\docplex-bin-avgonly.lp
..\WISER_Optimization_VG\data\1\31bonds\docplex-bin-minmaxavg.lp
..\WISER_Optimization_VG\data\1\31bonds\docplex.lp
..\WISER_Optimization_VG\data\1\full\docplex-full-reduced-nosemi.lp
..\WISER_Optimization_VG\data\1\full\docplex-orig.lp
..\WISER_Optimization_VG\data\1\gurobi\gurobi.lp


In [21]:
LP_PATH = "data/1/31bonds/docplex-bin-avgonly.lp"



In [ ]:
x, fx, dt = solve_bond_selection(LP_PATH)

print("🔢 Bonds selected:", int(np.sum(x)))
print("📈 Objective value:", round(fx, 4))
print("⏱ Runtime (seconds):", round(dt, 2))


In [ ]:
LP_PATH = "data/1/31bonds/docplex-bin-avgonly.lp"


In [ ]:
x, fx, dt = solve_bond_selection(LP_PATH)

print("🔢 Bonds selected:", int(np.sum(x)))
print("📈 Objective value:", round(fx, 4))
print("⏱ Runtime (seconds):", round(dt, 2))


In [ ]:
# --- Find/export LP (python-mip) OR use existing LP, then run tuned solve ---
from pathlib import Path
from datetime import datetime
import numpy as np

# 0) PROJECT_ROOT as Path
try:
    PROJECT_ROOT
except NameError:
    PROJECT_ROOT = Path.cwd()
else:
    PROJECT_ROOT = Path(PROJECT_ROOT)
print(f"[info] PROJECT_ROOT = {PROJECT_ROOT}")

# 1) Preferred targets and known LPs
results_lp = PROJECT_ROOT / "results" / "baseline.lp"
known_lp = PROJECT_ROOT / "data" / "1" / "31bonds" / "docplex-bin-avgonly.lp"

# 2) Try to detect a python-mip model (no need to know its variable name)
def find_mip_model():
    for name in ("model", "m", "mdl"):
        obj = globals().get(name)
        if obj and hasattr(obj, "write") and hasattr(obj, "optimize"):
            print(f"[info] Found python-mip model in variable: {name}")
            return obj
    # Try builder functions if they exist
    for builder in ("build_model", "make_model", "get_model", "create_model", "load_model", "build_mip_model"):
        fn = globals().get(builder)
        if callable(fn):
            try:
                print(f"[info] Building model via {builder}() ...")
                m = fn()
                if hasattr(m, "write") and hasattr(m, "optimize"):
                    print(f"[ok] Built python-mip model via {builder}()")
                    return m
            except Exception as e:
                print(f"[warn] {builder}() failed: {e}")
    return None

def export_lp_if_possible(target: Path) -> bool:
    m = find_mip_model()
    if not m:
        return False
    target.parent.mkdir(parents=True, exist_ok=True)
    try:
        m.write(str(target))
        ok = target.exists()
        print(f"[ok] Exported LP to: {target}" if ok else "[warn] Export failed")
        return ok
    except Exception as e:
        print(f"[warn] model.write('{target}') failed: {e}")
        return False

# 3) Choose an LP to solve
chosen = None

if results_lp.exists():
    chosen = results_lp
    print(f"[ok] Using: {chosen}")
elif known_lp.exists():
    chosen = known_lp
    print(f"[ok] Using known LP: {chosen}")
else:
    print("[info] No LP found at results/baseline.lp or data/1/31bonds/...")
    # Try exporting from in-memory model
    if export_lp_if_possible(results_lp):
        chosen = results_lp
    else:
        # Search project for any .lp
        matches = list(PROJECT_ROOT.rglob("*.lp"))
        if matches:
            # Prefer ones under data/ or results/
            preferred = [p for p in matches if any(part in ("data", "results") for part in p.parts)]
            chosen = (preferred[0] if preferred else matches[0])
            print("[warn] Falling back to discovered LP:")
            for i, p in enumerate(matches[:10], 1):
                print(f"  {i}. {p}")
            print(f"[use] Selected: {chosen}")
        else:
            raise FileNotFoundError(
                "No .lp file found and no python-mip model to export from.\n"
                "Options:\n"
                "  - Run your model-building cell so a python-mip Model exists, then re-run this cell.\n"
                "  - Place an LP file anywhere under the project (e.g., data/.../*.lp) and re-run.\n"
            )

# 4) Ensure we pass a RELATIVE path to solve_mip_tuned (it expects path relative to PROJECT_ROOT)
try:
    lp_rel = str(Path(chosen).resolve().relative_to(PROJECT_ROOT.resolve()))
except ValueError:
    # If outside project, copy name only as a last resort (or you can move the file under the project)
    lp_rel = str(Path(chosen).name)
    print(f"[warn] LP is outside PROJECT_ROOT. Using name only: {lp_rel}")

print(f"[run] Solving relative LP: {lp_rel}")
x2, fx2, dt2 = solve_mip_tuned(lp_rel, max_seconds=20, mip_gap=0.0)

# 5) Summary
def safe_sum(v):
    try:
        return int(np.sum(v))
    except Exception:
        try:
            return int(sum(v))
        except Exception:
            return None

summary = {
    "Timestamp": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
    "LP file": lp_rel,
    "Bonds selected": safe_sum(x2),
    "Objective value": float(fx2) if fx2 is not None else None,
    "Runtime (s)": round(float(dt2), 3) if dt2 is not None else None,
    "Solver": "python-mip",
    "Strategy": "tuned (20s, gap=0.0)",
}

print("— Results —")
print(f"  Bonds selected: {summary['Bonds selected']}")
print(f"  Objective:      {summary['Objective value']}")
print(f"  Runtime (s):    {summary['Runtime (s)']}")


In [ ]:
LP_REL = os.path.join(PROJECT_ROOT, "results", "baseline.lp")


In [ ]:
os.path.exists(LP_REL)


In [ ]:
# Step 1: Set LP file path (relative to project root)
LP_REL = "data/1/31bonds/docplex-bin-avgonly.lp"

# Sanity check the path
print("LP exists?", (PROJECT_ROOT / LP_REL).exists(), "->", (PROJECT_ROOT / LP_REL))

# Step 2: Run the solver (20s time limit)
x2, fx2, dt2 = solve_mip_tuned(LP_REL, max_seconds=20, mip_gap=0.0)

# Step 3: Print results
print("🔢 Bonds selected:", int(np.sum(x2)))
print("📈 Objective value:", round(fx2, 4))
print("⏱ Runtime (seconds):", round(dt2, 2))



In [ ]:
# Imports
import os, sys, time
from pathlib import Path
import numpy as np

# If python-mip isn't imported yet
from mip import Model

# Detect project root (works whether your notebook is in repo root or in notebooks/)
PROJECT_ROOT = Path.cwd()
if PROJECT_ROOT.name == "notebooks":
    PROJECT_ROOT = PROJECT_ROOT.parent
elif not (PROJECT_ROOT / "data").exists() and (PROJECT_ROOT.parent / "data").exists():
    PROJECT_ROOT = PROJECT_ROOT.parent

print("PROJECT_ROOT ->", PROJECT_ROOT)

def solve_mip_tuned(lp_rel_path, *, max_seconds=20, mip_gap=0.0, threads=0):
    """
    lp_rel_path: path relative to project root, e.g., 'data/1/31bonds/docplex-bin-avgonly.lp'
    max_seconds: time limit for the MIP solve
    mip_gap: allowed relative optimality gap (not all solvers honor this via API)
    threads: number of threads (0 lets solver decide)
    """
    lp_path = (PROJECT_ROOT / lp_rel_path).resolve()
    if not lp_path.exists():
        raise FileNotFoundError(f"LP file not found at: {lp_path}")

    start = time.perf_counter()
    m = Model()
    m.read(str(lp_path))

    # Optional knobs (wrapped in try/except to avoid solver-specific attribute issues)
    try:
        if threads:
            m.threads = threads
    except Exception:
        pass

    # Use the built-in time limit on optimize
    m.optimize(max_seconds=max_seconds)

    # Collect solution
    x = np.array([v.x for v in m.vars])
    fx = float(m.objective_value)
    dt = time.perf_counter() - start
    return x, fx, dt

print("✅ solve_mip_tuned is defined")


In [ ]:
LP_REL = "data/1/31bonds/docplex-bin-avgonly.lp"


In [ ]:
x2, fx2, dt2 = solve_mip_tuned(LP_REL, max_seconds=20, mip_gap=0.0)


In [ ]:
print("🔢 Bonds selected:", int(np.sum(x2)))
print("📈 Objective value:", round(fx2, 4))
print("⏱ Runtime (seconds):", round(dt2, 2))


In [ ]:
{
    "LP file": LP_REL,
    "Bonds selected": ...,
    "Objective value": ...,
    "Runtime (s)": ...,
    "Solver": "python-mip",
    "Strategy": "tuned (20s limit)"
}


In [ ]:
tuned_results = {
    "LP file": LP_REL,
    "Bonds selected": int(np.sum(x2)),
    "Objective value": round(fx2, 4),
    "Runtime (s)": round(dt2, 2),
    "Solver": "python-mip",
    "Strategy": "tuned (20s limit)"
}


In [ ]:
results_df = pd.DataFrame([tuned_results])
results_df


In [ ]:
import numpy as np
import pandas as pd

LP_REL = "data/1/31bonds/docplex-bin-avgonly.lp"


In [ ]:
baseline_results = {
    "LP file": LP_PATH,
    "Bonds selected": int(np.sum(x)),
    "Objective value": round(fx, 4),
    "Runtime (s)": round(dt, 2),
    "Solver": "python-mip",
    "Strategy": "docplex-bin-avgonly"
}

baseline_results


In [ ]:
import pandas as pd

results_df = pd.DataFrame([baseline_results])
results_df


### Tuned variant (20s limit)
This solver ran with a 20-second time limit, offering a faster alternative to the baseline.


## Solver Comparison Table

| Run | Solver         | LP File                   | Bonds | Objective | Runtime (s) | Notes          |
|-----|----------------|---------------------------|-------|-----------|-------------|----------------|
| 1   | Baseline (MIP) | docplex-bin-avgonly.lp    | 21    | 40.2771   | 139.88      | python-mip CLP |


In [ ]:
from pathlib import Path

CWD = Path.cwd()
print("CWD:", CWD)

lp_rel = Path("data/1/31bonds/docplex-bin-avgonly.lp")
lp_abs_guess = (CWD / lp_rel).resolve()
print("Expected LP path:", lp_abs_guess)
print("Exists?", lp_abs_guess.exists())


In [ ]:
from pathlib import Path

root = Path.cwd()
matches = list(root.rglob("docplex-bin-avgonly.lp"))
print("Matches found:", len(matches))
for i, p in enumerate(matches[:10], 1):
    print(f"{i}. {p}")

# Pick the first match if any were found
LP_ABS = matches[0] if matches else None
LP_ABS


In [ ]:
from pathlib import Path

PROJECT_ROOT = Path.cwd()
LP_ABS = LP_ABS.resolve()
print("Found LP file at:", LP_ABS)

# Try to compute a clean relative path (if it’s inside the project)
try:
    LP_REL = str(LP_ABS.relative_to(PROJECT_ROOT))
except ValueError:
    # If it’s outside the project, just use the absolute path
    LP_REL = str(LP_ABS)

print("LP_REL set to:", LP_REL)


In [ ]:
from pathlib import Path

path_to_open = Path(LP_REL)
if not path_to_open.is_absolute():
    path_to_open = (Path.cwd() / path_to_open).resolve()

print("Opening:", path_to_open)
with open(path_to_open, "r", encoding="utf-8", errors="ignore") as f:
    head = "".join([next(f) for _ in range(5)])
print("First lines:\n", head)


In [ ]:
LP_REL = str(lp_rel)
print("LP_REL set to:", LP_REL)


In [ ]:
from pathlib import Path

path_to_open = Path(LP_REL)
if not path_to_open.is_absolute():
    path_to_open = (Path.cwd() / path_to_open).resolve()

print("Opening:", path_to_open)
with open(path_to_open, "r", encoding="utf-8", errors="ignore") as f:
    head = "".join([next(f) for _ in range(5)])
print("First lines:\n", head)


In [ ]:
# Assuming solve_mip_tuned is already defined and imported in your notebook
x2, fx2, dt2 = solve_mip_tuned(LP_REL, max_seconds=20, mip_gap=0.0)

print("🔢 Bonds selected:", int(np.sum(x2)))
print("📈 Objective value:", round(fx2, 4))
print("⏱ Runtime (seconds):", round(dt2, 2))


In [ ]:
import os
print(os.getcwd())


In [ ]:
with open("../data/1/31bonds/docplex-bin-avgonly.lp", "r") as f:
    lp_contents = f.read()
